File: src/LabSync.Core/Entities/Device.cs
using LabSync.Core.Types;
using System.ComponentModel.DataAnnotations;

namespace LabSync.Core.Entities;

public class Device
{
    public Guid Id { get; init; }
    public string Hostname { get; private set; }
    public bool IsApproved { get; private set; }
    public string MacAddress { get; init; }
    public string? IpAddress { get; private set; }
    public DevicePlatform Platform { get; private set; }
    public string OsVersion { get; private set; }
    public DeviceStatus Status { get; private set; }
    public bool IsOnline { get; private set; }
    public DateTime RegisteredAt { get; init; }
    public DateTime? LastSeenAt { get; private set; }

    public Guid? GroupId { get; private set; }
    public DeviceGroup? Group { get; private set; }

    public string DeviceKeyHash { get; private set; }

    private readonly List<Job> _jobs = new();
    public IReadOnlyCollection<Job> Jobs => _jobs.AsReadOnly();

    protected Device() { }

    public Device(string hostname, string macAddress, DevicePlatform platform, string osVersion, string deviceKeyHash)
    {
        Id = Guid.NewGuid();
        Hostname      = hostname;
        MacAddress    = macAddress;
        Platform      = platform;
        OsVersion     = osVersion;
        DeviceKeyHash = deviceKeyHash;
        RegisteredAt  = DateTime.UtcNow;
        Status        = DeviceStatus.Pending;
        IsApproved    = false;
        IsOnline      = false;
    }

    public void Approve()
    {
        IsApproved = true;
        Status     = DeviceStatus.Active;
    }

    public void MarkAsOffline()
    {
        IsOnline = false;
    }

    public void RecordHeartbeat(string ipAddress)
    {
        IsOnline   = true;
        IpAddress  = ipAddress;
        LastSeenAt = DateTime.UtcNow;
    }
}

File: src/client/src/types/device.ts
/**
 * Mirrors LabSync.Core.ValueObjects.DevicePlatform
 */
export const DevicePlatform = {
  Unknown: 0,
  Windows: 1,
  Linux: 2,
  MacOS: 3,
} as const;

export type DevicePlatform = (typeof DevicePlatform)[keyof typeof DevicePlatform];

/**
 * Mirrors LabSync.Core.ValueObjects.DeviceStatus
 */
export const DeviceStatus = {
  Pending: 0,
  Active: 1,
  Maintenance: 2,
  Blocked: 3,
} as const;

export type DeviceStatus = (typeof DeviceStatus)[keyof typeof DeviceStatus];

/**
 * Mirrors LabSync.Core.Dto.DeviceDto
 */
export interface DeviceDto {
  id: string;
  hostname: string;
  isApproved: boolean;
  macAddress: string;
  ipAddress: string | null;
  platform: DevicePlatform;
  osVersion: string;
  status: DeviceStatus;
  isOnline: boolean;
  registeredAt: string;
  lastSeenAt: string | null;
  groupId: string | null;
  groupName: string | null;
  hardwareInfo: string | null;
}

/**
 * Mirrors LabSync.Core.Dto.ApiResponse
 */
export interface ApiResponse {
  message: string;
}

export const DEVICE_PLATFORM_LABELS: Record<DevicePlatform, string> = {
  [DevicePlatform.Unknown]: 'Unknown',
  [DevicePlatform.Windows]: 'Windows',
  [DevicePlatform.Linux]: 'Linux',
  [DevicePlatform.MacOS]: 'MacOS',
};

export const DEVICE_STATUS_LABELS: Record<DeviceStatus, string> = {
  [DeviceStatus.Pending]: 'Pending',
  [DeviceStatus.Active]: 'Active',
  [DeviceStatus.Maintenance]: 'Maintenance',
  [DeviceStatus.Blocked]: 'Blocked',
};
